### PR TITLE
Final part of the binary targets implementation

### PR DESCRIPTION
**Pull request by hartbit**
The process for linking against the XCFrameworks mimics how Xcode does it:

* Copying the correct `.framework` in the build directory,
* Adding a `-F` option to the compiler arguments pointing to the build directory,
* Adding a `-F` option to the linker arguments pointing to the build directory.
* Adding a `-framework` option to the linker arguments with the name of the framework.

Based on https://github.com/apple/swift-package-manager/pull/2511

### CODE REVIEW

**Comment by neonichu** [comment]
In addition to frameworks, XCFrameworks can also contain a dynamic/static library + headers. We should handle that case here as well.

**Comment by hartbit** [comment]
Thanks, I had forgotten about that case.

**Comment by hartbit** [comment]
I now handle static and dynamic libraries.

File: Fixtures/BinaryTargets/Framework.xcframework/macos-x86_64/Framework.framework/Versions/A/Headers/Framework.h [hunk 1]
```diff
@@ -0,0 +1,19 @@
+//
+//  Framework.h
+//  Framework
+//
+//  Created by David Hart on 14.01.20.
+//  Copyright © 2020 David Hart. All rights reserved.
```
Comment (aciidb0mb3r) [inline Fixtures/BinaryTargets/Framework.xcframework/macos-x86_64/Framework.framework/Versions/A/Headers/Framework.h]: Can you change this to Swift.org copyright headers instead?

File: Fixtures/BinaryTargets/Framework.xcframework/macos-x86_64/Framework.framework/Versions/A/Modules/Framework.swiftmodule/x86_64.swiftinterface [hunk 2]
```diff
@@ -0,0 +1,9 @@
+// swift-interface-format-version: 1.0
```
Comment (aciidb0mb3r) [inline Fixtures/BinaryTargets/Framework.xcframework/macos-x86_64/Framework.framework/Versions/A/Modules/Framework.swiftmodule/x86_64.swiftinterface]: We probably don't need the above swiftdoc binary file.

File: Sources/Build/BuildPlan.swift [hunk 3]
```diff
@@ -247,6 +260,8 @@ public final class ClangTargetBuildDescription {
             // Using modules currently conflicts with the Windows and Android SDKs.
             args += ["-fmodules", "-fmodule-name=" + target.c99name]
         }
+
+        args += ["-F", buildParameters.buildPath.pathString]
```
Comment (aciidb0mb3r) [inline Sources/Build/BuildPlan.swift]: can you make this conditional on presence of at least one framework?

File: Sources/Build/BuildPlan.swift [hunk 4]
```diff
@@ -544,6 +562,8 @@ public final class SwiftTargetBuildDescription {
         args += moduleCacheArgs
         args += buildParameters.sanitizers.compileSwiftFlags()
         args += ["-parseable-output"]
+        args += ["-F", buildParameters.buildPath.pathString]
+        args += ["-L", buildParameters.buildPath.pathString]
```
Comment (aciidb0mb3r) [inline Sources/Build/BuildPlan.swift]: This shouldn't be required since we don't link here.

File: Sources/Build/BuildPlan.swift [hunk 5]
```diff
@@ -544,6 +562,8 @@ public final class SwiftTargetBuildDescription {
         args += moduleCacheArgs
         args += buildParameters.sanitizers.compileSwiftFlags()
         args += ["-parseable-output"]
+        args += ["-F", buildParameters.buildPath.pathString]
```
Comment (aciidb0mb3r) [inline Sources/Build/BuildPlan.swift]: can you make this conditional on presence of at least one framework?

File: Sources/Build/BuildPlan.swift [hunk 6]
```diff
@@ -935,6 +955,7 @@ public final class ProductBuildDescription {
             }
         }
 
+        args += ["-F", buildParameters.buildPath.pathString]
```
Comment (aciidb0mb3r) [inline Sources/Build/BuildPlan.swift]: can you make this conditional on presence of at least one framework?

**Comment by aciidb0mb3r** [comment]
Some minor comments but otherwise LGTM

File: Fixtures/BinaryTargets/Framework.xcframework/macos-x86_64/Framework.framework/Versions/A/Headers/Framework.h (hunk 1 above)
Comment (hartbit) [reply Fixtures/BinaryTargets/Framework.xcframework/macos-x86_64/Framework.framework/Versions/A/Headers/Framework.h]: Fixed in future commit.

File: Fixtures/BinaryTargets/Framework.xcframework/macos-x86_64/Framework.framework/Versions/A/Modules/Framework.swiftmodule/x86_64.swiftinterface (hunk 2 above)
Comment (hartbit) [reply Fixtures/BinaryTargets/Framework.xcframework/macos-x86_64/Framework.framework/Versions/A/Modules/Framework.swiftmodule/x86_64.swiftinterface]: I just generated an XCFramework using Xcode and the standard procedure. Do you want me to remove just this file?

File: Sources/Build/BuildPlan.swift (hunk 3 above)
Comment (hartbit) [reply Sources/Build/BuildPlan.swift]: I did it this way to mimic Xcode, which always includes it, even when there are no frameworks to link against. Is there a reason they do it that way?

File: Fixtures/BinaryTargets/Framework.xcframework/macos-x86_64/Framework.framework/Versions/A/Modules/Framework.swiftmodule/x86_64.swiftinterface (hunk 2 above)
Comment (aciidb0mb3r) [reply Fixtures/BinaryTargets/Framework.xcframework/macos-x86_64/Framework.framework/Versions/A/Modules/Framework.swiftmodule/x86_64.swiftinterface]: Yeah. I don't think we really need it.

Comment (hartbit) [reply Fixtures/BinaryTargets/Framework.xcframework/macos-x86_64/Framework.framework/Versions/A/Modules/Framework.swiftmodule/x86_64.swiftinterface]: Done in next commit.

File: Sources/Build/BuildPlan.swift (hunk 4 above)
Comment (hartbit) [reply Sources/Build/BuildPlan.swift]: I can't remember why I added it, but it does seem un-necessary indeed.

File: Fixtures/BinaryTargets/Framework.xcframework/Info.plist [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+<plist version="1.0">
+<dict>
+	<key>AvailableLibraries</key>
+	<array>
+		<dict>
+			<key>LibraryIdentifier</key>
+			<string>macos-x86_64</string>
+			<key>LibraryPath</key>
+			<string>Framework.framework</string>
```
Comment (aciidb0mb3r) [inline Fixtures/BinaryTargets/Framework.xcframework/Info.plist]: Can you call it something else? `Framework.framework` is a bit confusing 😁

File: Sources/Build/BuildPlan.swift (hunk 3 above)
Comment (hartbit) [reply Sources/Build/BuildPlan.swift]: Done in latest commit.

File: Fixtures/BinaryTargets/Framework.xcframework/Info.plist (hunk 7 above)
Comment (hartbit) [reply Fixtures/BinaryTargets/Framework.xcframework/Info.plist]: I called it that way to clearly differentiate it from the static and dynamic library binaries. Do you have a suggestion?

Comment (aciidb0mb3r) [reply Fixtures/BinaryTargets/Framework.xcframework/Info.plist]: `MyFwk`?

**Comment by hartbit** [comment]
@swift-ci smoke test

**Comment by hartbit** [comment]
@swift-ci smoke test

**Comment by hartbit** [comment]
@swift-ci smoke test

**Pull request merged**